### PR TITLE
RegexPrefix struct, RegexFCD buffer to ValueListBuilder & ArrayPool

### PR DESCRIPTION
**Pull request by ViktorHofer**
**Updated results from yesterday (couldn't repro the better number in commit 2).**

before:

|     Method |     Mean |    Error |    StdDev |       Gen 0 | Allocated |
|----------- |---------:|---------:|----------:|------------:|----------:|
|  RegexCtor | 339.5 ms | 0.9025 ms | 0.8442 ms | 201000.0000 | 402.07 MB |

after:

|     Method |     Mean |    Error |    StdDev |       Gen 0 | Allocated |
|----------- |---------:|---------:|----------:|------------:|----------:|
|  RegexCtor | 312.0 ms | 1.447 ms | 1.1258 ms | 189125.0000 | 378.36 MB |

### Results

After the first regex construction:
~ 8% increased throughput
~ 4% less allocation

### CODE REVIEW

File: src/System.Text.RegularExpressions/src/System.Text.RegularExpressions.csproj [hunk 1]
```diff
@@ -47,13 +53,16 @@
...
+      <Link>Common\System\Collections\HashtableExtensions.cs</Link>
+    </Compile>
   </ItemGroup>
   <!-- Files that enable compiled feature -->
   <ItemGroup Condition="'$(TargetGroup)' == 'netcoreapp'">
-    <Compile Include="System\Text\RegularExpressions\CompiledRegexRunnerFactory.cs" />
-    <Compile Include="System\Text\RegularExpressions\CompiledRegexRunner.cs" />
-    <Compile Include="System\Text\RegularExpressions\RegexCompiler.cs" />
-    <Compile Include="System\Text\RegularExpressions\RegexLWCGCompiler.cs" />
+    <Compile Include="System\Text\RegularExpressions\Runner\CompiledRegexRunnerFactory.cs" />
```
Comment (jkotas) [inline src/System.Text.RegularExpressions/src/System.Text.RegularExpressions.csproj]: Coding conventions says that directory should match the namespace. There is no System.Text.RegularExpressions.Runner namespace...

File: src/System.Text.RegularExpressions/src/System.Text.RegularExpressions.csproj [hunk 2]
```diff
@@ -47,13 +53,16 @@
     <Compile Include="$(CommonPath)\CoreLib\System\Collections\Generic\ValueListBuilder.cs">
       <Link>Common\System\Collections\Generic\ValueListBuilder.cs</Link>
     </Compile>
+    <Compile Include="$(CommonPath)\CoreLib\System\Collections\HashtableExtensions.cs">
```
Comment (jkotas) [inline src/System.Text.RegularExpressions/src/System.Text.RegularExpressions.csproj]: It is not used by anything else. Why move it to common?

File: src/System.Text.RegularExpressions/src/System.Text.RegularExpressions.csproj (hunk 1 above)
Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/src/System.Text.RegularExpressions.csproj]: I didn't know that we are enforcing that. Folders especially here make sense (as we don't have an sub namespaces) but I'm fine with flattening it back.

**Comment by ViktorHofer** [comment]
Thanks Jan. Updated the PR. Also cc @stephentoub for perf

**Comment by danmosemsft** [comment]
YOu might grab the comments in https://github.com/dotnet/corefx/compare/master...danmosemsft:regex.commenting?expand=1 also.

File: src/System.Text.RegularExpressions/src/System.Text.RegularExpressions.csproj [hunk 3]
```diff
@@ -12,31 +12,38 @@
   <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'uap-Windows_NT-Debug|AnyCPU'" />
   <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'uap-Windows_NT-Release|AnyCPU'" />
   <ItemGroup>
-    <Compile Include="System\Text\RegularExpressions\HashtableExtensions.cs" />
+    <Compile Include="System\Collections\Generic\ValueListBuilder.Pop.cs" />
```
Comment (danmosemsft) [inline src/System.Text.RegularExpressions/src/System.Text.RegularExpressions.csproj]: This should probably merge into the rest of VLB.

File: src/System.Text.RegularExpressions/src/System.Text.RegularExpressions.csproj [hunk 4]
```diff
@@ -12,31 +12,38 @@
...
+    <Compile Include="System\Text\RegularExpressions\Regex.Timeout.cs" />
     <Compile Include="System\Text\RegularExpressions\Regex.cs" />
     <Compile Include="System\Text\RegularExpressions\RegexBoyerMoore.cs" />
-    <Compile Include="System\Text\RegularExpressions\RegexCapture.cs" />
-    <Compile Include="System\Text\RegularExpressions\RegexCaptureCollection.cs" />
     <Compile Include="System\Text\RegularExpressions\RegexCharClass.cs" />
     <Compile Include="System\Text\RegularExpressions\RegexCode.cs" />
-    <Compile Include="System\Text\RegularExpressions\RegexCollectionDebuggerProxy.cs" />
-    <Compile Include="System\Text\RegularExpressions\RegexCompilationInfo.cs" />
     <Compile Include="System\Text\RegularExpressions\RegexFCD.cs" />
```
Comment (danmosemsft) [inline src/System.Text.RegularExpressions/src/System.Text.RegularExpressions.csproj]: Did you figure out what FCD stands for? Probably First Child D...  ?

File: src/System.Text.RegularExpressions/src/System.Text.RegularExpressions.csproj (hunk 3 above)
Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/src/System.Text.RegularExpressions.csproj]: Stephen advised to keep the Pop method out

File: src/System.Text.RegularExpressions/src/System.Text.RegularExpressions.csproj (hunk 4 above)
Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/src/System.Text.RegularExpressions.csproj]: hmm. not really. comments in its file:

// This RegexFCD class is internal to the Regex package.
// It builds a bunch of FC information (RegexFC) about

File: src/System.Text.RegularExpressions/src/System.Text.RegularExpressions.csproj (hunk 3 above)
Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/src/System.Text.RegularExpressions.csproj]: Hmm I believe it was Stephen, or Jan. But that was before we had the Indexer in it. I don't have a strong opinion here...

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs [hunk 5]
```diff
@@ -472,15 +471,16 @@ internal static string AnchorDescription(int anchors)
 #endif
     }
 
-    internal sealed class RegexFC
+    internal struct RegexFC : IDisposable
```
Comment (danmosemsft) [inline src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs]: IDisposable may not be a good fit for a struct especially when forgetting to dispose is not a resourcel eak.

https://github.com/dotnet/corefx/pull/27250#discussion_r169745944

Comment (danmosemsft) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs]: You are not using it either you are calling Dispose directly.

**Comment by ViktorHofer** [comment]
Updated with RegexFC ArrayPool usage.

With our test data the RegexFC's buffer is always < 32 which means it never grows. As it was implemented as a growing stack I did the same here. I looked at the code but am not 100% sure if it could ever get larger than its initial size. Not really important here.

The more important change is that I now rent RegexFC (struct) arrays from ArrayPool. I don't know if we pass a managed type to ArrayPool anywhere else in Core but after talking with Stephen I believe the change is safe. To avoid memory leaks I clear the reference to RegexCharClass when disposing.

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs [hunk 6]
```diff
@@ -58,19 +58,26 @@ private RegexFCD(ValueListBuilder<int> intStack)
             // Create/rent buffers
             Span<int> intSpan = stackalloc int[StackBufferSize];
             var intStack = new ValueListBuilder<int>(intSpan);
-
-            RegexFCD s = new RegexFCD(intStack);
-            RegexFC fc = s.RegexFCFromRegexTree(t);
-
-            // Return rented buffers
+            RegexFC[] fcSpan = ArrayPool<RegexFC>.Shared.Rent(StackBufferSize);
+            var fcStack = new ValueListBuilder<RegexFC>(fcSpan, clearArray: true);
```
Comment (danmosemsft) [inline src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs]: why is clearArray important? we almost never clear on returning.

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs (hunk 5 above)
Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs]: it's not a ref type but I agree, we don't need it here. I had a different implementation for ValueListBuilder which detects if the type implements IDisposable and then calls it but there was a perf regression with the type is IDisposable check (reflection).

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs (hunk 6 above)
Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs]: the clearArray here is wrong as I do it manually some lines below but in general clearArray is important if you store managed objects with references to other objects (in our case RegexFC --> RegexCharClass). We usually use ArrayPool in combination with primitive types like int & char.

Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs]: I accidently merged two commits, don't wonder :)

**Comment by danmosemsft** [comment]
OK so far but please please next time you do a big refactoring keep each refactoring into its own commit. I don't care how small the commits are. Eg., move A to B, rename C to D, change E to struct, split F into partial classes G and H. The split into partial classes is particularly difficult to review because I would have to manually set up the diff between old and new files. I did not see anything much but doing it like that makes it possible to be much more confident and do the review FAR more quickly and accurately.

Ideally you would have the refactoring first and behavior changes in separate PR's. Since you've started this though, let's at least leave them in separate commits (ie, rebase not squash) so that we have some hope of bisecting if a problem emerges.

When refactoring good test coverage is critical. As we noted offline the block code coverage of regex is very good, much of the unhit blocks are debug only code. Of course by the nature of regexes, block coverage is hardly sufficient, but it could be worse.

**Comment by danmosemsft** [comment]
Great to see the improvements though!

**Comment by ViktorHofer** [comment]
Thanks Dan, I appreciate the feedback (as discussed offline) and will try to avoid this behavior in future. I will now continue with the next candidate and will open a new PR after this one is **merged (not squashed)** into master.

File: src/System.Text.RegularExpressions/src/System.Text.RegularExpressions.csproj (hunk 3 above)
Comment (stephentoub) [reply src/System.Text.RegularExpressions/src/System.Text.RegularExpressions.csproj]: IMHO, Pop as an operation doesn't really belong on the type; it's just how Regex happens to use it.  My original suggestion had actually been (or what I'd intended) for Regex to just have a static helper that took a ValueListBuilder by ref, rather than having it be a partial class instance method like this.

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Capture.cs [hunk 7]
```diff
@@ -15,91 +15,63 @@ namespace System.Text.RegularExpressions
...
+        /// The substring to the right of the capture
         /// </summary>
-        public int Length
+        internal string GetRightSubstring()
         {
-            get
-            {
-                return _length;
-            }
+            return Text.Substring(Index + Length, Text.Length - Index - Length);
```
Comment (stephentoub) [inline src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Capture.cs]: How about slicing rather than substringing?  It looks like at least one of the callers then appends the result to a StringBuilder, so if you do it as a slice, you can avoid the intermediary string.

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Capture.cs [hunk 8]
```diff
@@ -15,91 +15,63 @@ namespace System.Text.RegularExpressions
...
         /// </summary>
-        public int Index
+        /// <returns></returns>
+        internal string GetLeftSubstring()
         {
-            get
-            {
-                return _index;
-            }
+            return Text.Substring(0, Index);
```
Comment (stephentoub) [inline src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Capture.cs]: Same slicing comment/question.

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Match.cs [hunk 9]
```diff
@@ -259,25 +202,43 @@ internal virtual void BalanceMatch(int cap)
...
-         */
+        internal virtual string GroupToStringImpl(int groupnum)
+        {
+            int c = _matchcount[groupnum];
+            if (c == 0)
+                return string.Empty;
+
+            int[] matches = _matches[groupnum];
+
+            return Text.Substring(matches[(c - 1) * 2], matches[(c * 2) - 1]);
```
Comment (stephentoub) [inline src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Match.cs]: Same slicing comment/question.

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Reference.cs [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+using System.Threading;
+
+namespace System.Text.RegularExpressions
+{
+    /// <summary>
+    /// Used to cache a weak reference in a threadsafe way
+    /// </summary>
+    internal sealed class SharedReference
+    {
+        private WeakReference _ref = new WeakReference(null);
```
Comment (stephentoub) [inline src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Reference.cs]: readonly

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Reference.cs [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+        private int _locked;
```
Comment (stephentoub) [inline src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Reference.cs]: _locked should be volatile

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Reference.cs [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+    }
+
+    /// <summary>
+    /// Used to cache one exclusive runner reference
+    /// </summary>
+    internal sealed class ExclusiveReference
+    {
+        private RegexRunner _ref;
+        private object _obj;
+        private int _locked;
```
Comment (stephentoub) [inline src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Reference.cs]: _locked should be volatile

**Comment by ViktorHofer** [comment]
@stephentoub 

The first commit is intended to just do refactoring / code styling stuff. The 2nd and 3rd commits are about performance of the RegexFCD and RegexFC types and about making RegexPrefix a struct.

I touched a lot of files in the first commit but intend to work on them in separate PRs as there are tons of possible improvements there. E.g. the slicing you pointed out.

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Reference.cs [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+        /// Return an object and grab an exclusive lock.
+        /// 
+        /// If the exclusive lock can't be obtained, null is returned.
+        /// If the object can't be returned, the lock is released.
+        /// </summary>
+        /// <returns></returns>
+        internal object Get()
+        {
+            // try to obtain the lock
+            if (0 == Interlocked.Exchange(ref _locked, 1))
```
Comment (stephentoub) [inline src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Reference.cs]: Why not just use a normal lock?

**Comment by ViktorHofer** [comment]
Thanks Stephen, I will incorporate your comments in my next PR which will bring the refactorings back. If you want you can take a look at my (mostly) perf changes here.

cc @danmosemsft PTAL

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs [hunk 14]
```diff
@@ -1453,27 +1453,27 @@ internal void GenerateFindFirstChar()
                 else
                     Rightcharnext();
 
-                if (_fcPrefix.CaseInsensitive)
+                if (_fcPrefix.Value.CaseInsensitive)
```
Comment (stephentoub) [inline src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs]: Do we know at all of these places where Value is being accessed that it's non-null?  If so, GetValueOrDefault() is slightly faster, as it doesn't involve a null check, whereas Value does (throwing if it's null).

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs [hunk 15]
```diff
@@ -39,20 +34,50 @@ internal sealed class RegexFCD
...
+            RegexFC[] fcSpan = ArrayPool<RegexFC>.Shared.Rent(StackBufferSize);
+            var fcStack = new ValueListBuilder<RegexFC>(fcSpan);
+
+            RegexFCD s = new RegexFCD(fcStack, intStack);
+            RegexFC? fc = s.RegexFCFromRegexTree(t);
+
+            // Return rented buffers. Clear RegexFC buffer manually as only
+            // the reference to the RegexCharClass needs to be cleared.
+            for (int i = 0; i < fcStack.Length; i++)
+                fcStack[i].Dispose();
```
Comment (stephentoub) [inline src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs]: Each of these fcStacks has a rented buffer?  How many of them are there?  If we're consuming lots of ArrayPool buffers for one regex operation, what happens when regexs are performed in parallel, such as in an ASP.NET application with lots of requests all being processed concurrently?

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs [hunk 16]
```diff
@@ -39,20 +34,50 @@ internal sealed class RegexFCD
...
+        /// <summary>
+        /// This is the one of the only two functions that should be called from outside.
+        /// It takes a RegexTree and computes the set of chars that can start it.
+        /// </summary>
+        internal static RegexPrefix? FirstChars(RegexTree t)
+        {
+            // Create/rent buffers
+            Span<int> intSpan = stackalloc int[StackBufferSize];
+            var intStack = new ValueListBuilder<int>(intSpan);
+            RegexFC[] fcSpan = ArrayPool<RegexFC>.Shared.Rent(StackBufferSize);
```
Comment (stephentoub) [inline src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs]: We're renting an array of StackBufferSize?  I'd have thought we'd start with stack space of StackBufferSize.  I guess we can't as these aren't primitives?

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs (hunk 15 above)
Comment (stephentoub) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs]: Nevermind, I misunderstood, and I see further on that they don't wrap rented buffers.

Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs]: still a good question! That code is only called once per Regex construction, it's part of the pre-processing stuff (tree, opcode generation, ...). Of course, if multiple instances of Regex are generated while one of the arrays is rented and not yet returned the ArrayPool will allocate a new array with size 32. Not get rid off the unused memory after time @jkotas thinks about integrating the ArrayPool with the GC, described here: https://github.com/dotnet/coreclr/issues/16502#issuecomment-367720105

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs (hunk 16 above)
Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs]: correct. it's a managed type which can't be stackalloced. I don't know why these restriction apply even to ref structs that always live on the stack. maybe @AndyAyersMS can answer?

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs (hunk 14 above)
Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs]: yes, everywhere where I added `Value.` it's safe to assume that the value is never null. Thanks, didn't know that.

File: src/Common/src/CoreLib/System/Collections/Generic/ValueListBuilder.cs [hunk 17]
```diff
@@ -13,12 +13,14 @@ namespace System.Collections.Generic
         private Span<T> _span;
         private T[] _arrayFromPool;
         private int _pos;
+        private readonly bool _clearArray;
```
Comment (jkotas) [inline src/Common/src/CoreLib/System/Collections/Generic/ValueListBuilder.cs]: We should not be making ValueListBuilder more expensive by adding cruft like this to it.

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/ValueListBuilder.Pop.cs [hunk 18]
```diff
@@ -12,10 +12,10 @@ namespace System.Collections.Generic
     internal ref partial struct ValueListBuilder<T>
     {
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public T Pop()
+        public ref T Pop()
```
Comment (jkotas) [inline src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/ValueListBuilder.Pop.cs]: This is pretty dangerous. Very easy to make a mistake by using the value after it has been overwritten.

Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/ValueListBuilder.Pop.cs]: I'm not using the Pop function with the ref keyword, I just wanted to add the possibility to do so. I understand the concerns.

File: src/Common/src/CoreLib/System/Collections/Generic/ValueListBuilder.cs (hunk 17 above)
Comment (ViktorHofer) [reply src/Common/src/CoreLib/System/Collections/Generic/ValueListBuilder.cs]: As I'm not using these changes here I will revert them.

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/ValueListBuilder.Pop.cs (hunk 18 above)
Comment (jkotas) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/ValueListBuilder.Pop.cs]: If you do not actually needs this, this change should be undone.

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs (hunk 16 above)
Comment (jkotas) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs]: Right, stackallocation works for primitive/blitable types only.

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs (hunk 15 above)
Comment (jkotas) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs]: Have you measured that you are really getting any speedup from using pooled arrays for these? The GC is very good at getting rid of small short-term garbage that this is producing. I think that we are going overboard with trying to pool here.

File: src/Common/src/CoreLib/System/Collections/Generic/ValueListBuilder.cs (hunk 17 above)
Comment (ViktorHofer) [reply src/Common/src/CoreLib/System/Collections/Generic/ValueListBuilder.cs]: I already reverted the change but I wouldn't call that addition cruft. In cases where you want to use ValueListBuilder with types that need to cleaned up, the clearArray overload is handy. In my case I didn't use it as it is faster to partially clean it up instead of using Array.Clear.

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs (hunk 15 above)
Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs]: Throughput stays same, allocation reduction is the only benefit by using ArrayPools for RegexFCD. With commit 1 I was using a list with an initial size of 32 which results in same throughput but 40MB more allocation. I don't have a strong opinion here and I see that change more as an experiment. I was talking with Stephen about it offline beforehand and we concluded that it is worth giving it a try. As said, I'm absolutely fine with using a List here for the RegexFCD data store.

Comment (jkotas) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs]: Then I would not be pooling here. I think allocations during regex construction are ok, it is not something that we should go out of our way to optimize.

File: src/Common/src/CoreLib/System/Collections/Generic/ValueListBuilder.cs (hunk 17 above)
Comment (danmosemsft) [reply src/Common/src/CoreLib/System/Collections/Generic/ValueListBuilder.cs]: Pint is it's not pay for play though, it makes everyone use more stack

Comment (jkotas) [reply src/Common/src/CoreLib/System/Collections/Generic/ValueListBuilder.cs]: You cannot really build common helper classes like these to support everything that one might potentially want to do. It kills their value. They need to stay focused on the mainstream use.

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs (hunk 16 above)
Comment (AndyAyersMS) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs]: Stackalloc storage can't contain gc refs. Among other reasons, there is no mechanism for gc reporting from stackalloc buffers.

On the surface, the idea of allowing GC refs in a stackalloc buffer seems viable, if we can get sufficient safety guarantees. Say if the stackalloc is zero-init and immediately span wrapped and there is no safe way to view it as raw memory. With the right implementation (say a jit intrisnic Span ctor) the jit could know the type of the elements and create a new kind of gc info report for it. But we'd have to look into it in more detail to see if other problems crop up.

Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs]: Sounds interesting. Andy could you please open an issue for that? I think you could describe the matter far better than me.

**Comment by ViktorHofer** [comment]
Thanks, I addressed feedback. PTAL/approve for merging.

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs (hunk 16 above)
Comment (jkotas) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs]: https://github.com/dotnet/csharplang/blob/master/proposals/fixed-sized-buffers.md and related issues are pretty much this. @VSadov has been working on this recently.

Comment (VSadov) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs]: There is a proposal and a prototype where you could just say
```C#
RegexFC[StackBufferSize] fcSpan = default; 

... just use fcSpan  as an indexable array-like struct ...

```

---
In reply to: [170391575](https://github.com/dotnet/corefx/pull/27314#discussion_r170391575) [](ancestors = 170391575)

Comment (VSadov) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs]: It could be useful in scenarios like this, but for now, I guess, array renting is a solution.

---
In reply to: [170393377](https://github.com/dotnet/corefx/pull/27314#discussion_r170393377) [](ancestors = 170393377,170391575)

Comment (AndyAyersMS) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs]: Supporting GC refs in span-wrapped stackallocs is similar to the fixed-sized buffer proposal but also seems slightly different, in perhaps interesting ways (?). It would allow for fixed or variable sized buffers and would be stack only, instead of supporting only fixed sized buffers but being storage agnostic. Would not require any language changes, just a new magic span constructor that the jit turns into stackalloc under the covers...

Comment (VSadov) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs]: Yes. The fixed size arrays are a syntactical sugar for ordinary struct types of appropriate size. You can have locals, parameters, fields of those types. They can even nest.
GC-tracked stackalloc allows for variable number of elements, but values must be stack-only. That would be basically the tradeoff.

---
In reply to: [170400977](https://github.com/dotnet/corefx/pull/27314#discussion_r170400977) [](ancestors = 170400977)

**Comment by ViktorHofer** [comment]
fixed failing tests

**Comment by ViktorHofer** [comment]
@dotnet-bot test Linux x64 Release Build
@dotnet-bot test OSX x64 Debug Build

**Comment by ViktorHofer** [comment]
@danmosemsft @stephentoub I think this PR is now ready to be merged. Please review/approve.

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Capture.cs (hunk 7 above)
Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Capture.cs]: Are you talking about Span slicing here?

Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Capture.cs]: Got it. I think you meant something like this:

`internal string GetLeftSubstring() => Text.AsReadOnlySpan(0, Index).ToString();`

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Reference.cs (hunk 13 above)
Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Reference.cs]: No idea. I think that change was made by @justinvp. Maybe he can answer?

Comment (justinvp) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Reference.cs]: I didn't change it. This is how it was when the sources were added to corefx.

https://github.com/dotnet/corefx/commit/cc07fbf1a2fdac12523573ffb2dd273d82542f8d#diff-18063ba409511b10aec57135398e72d3R1056

**Comment by ViktorHofer** [comment]
@dotnet/dnceng Not the first time that I notice that legs mentioned in the previous comment restart when posting a new comment.

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Reference.cs (hunk 13 above)
Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Reference.cs]: thanks and sorry for the noise!

**Comment by jonfortescue** [comment]
@ViktorHofer Huh. That seems strange. We'll look into it.

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs [hunk 19]
```diff
@@ -210,14 +219,22 @@ public static string AnchorDescription(int anchors)
...
 
-            if (0 != (anchors & Beginning)) sb.Append(", Beginning");
-            if (0 != (anchors & Start)) sb.Append(", Start");
-            if (0 != (anchors & Bol)) sb.Append(", Bol");
-            if (0 != (anchors & Boundary)) sb.Append(", Boundary");
-            if (0 != (anchors & ECMABoundary)) sb.Append(", ECMABoundary");
-            if (0 != (anchors & Eol)) sb.Append(", Eol");
-            if (0 != (anchors & End)) sb.Append(", End");
-            if (0 != (anchors & EndZ)) sb.Append(", EndZ");
+            if (0 != (anchors & Beginning))
```
Comment (ViktorHofer) [inline src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs]: VS did that change...

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs [hunk 20]
```diff
@@ -400,27 +359,27 @@ private void CalculateFC(int NodeType, RegexNode node, int CurIndex)
...
+                    _fcStack.Add(new RegexFC(true));
                     break;
 
                 case RegexNode.Concatenate | AfterChild:
                     if (CurIndex != 0)
                     {
-                        RegexFC child = PopFC();
-                        RegexFC cumul = TopFC();
+                        RegexFC child = FCPop();
+                        RegexFC cumul = _fcStack[_fcStack.Count - 1];
```
Comment (stephentoub) [inline src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs]: Why did you change all of these away from the helper method?

Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs]: When I was trying a different data model I probably removed those. When I prepared the PR today for review today I noticed that I should revert these changes but was too lazy. Will do that now ;)

**Comment by ViktorHofer** [comment]
Thanks Stephen! Yes I have some parallel tests locally but we don't have any in corefx yet. I will create an issue to track that.

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs [hunk 21]
```diff
@@ -226,79 +242,57 @@ public static string AnchorDescription(int anchors)
...
-         * This is the top.
-         */
         private RegexFC TopFC()
         {
-            return _fcStack[_fcDepth - 1];
+            return _fcStack[_fcStack.Count - 1];
+        }
+
+        /// <summary>
+        /// Return rented buffers. Clear RegexFC buffer manually as only
```
Comment (stephentoub) [inline src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs]: What code is the "Clear RegexFC buffer manually" referring to?

Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFCD.cs]: thanks.
